src/main/java/com/google/api/codegen/ApiaryConfig.java

```diff
@@ -102,15 +102,59 @@
   private final Table<Type, String, Field> fields = HashBasedTable.<Type, String, Field>create();
 
   /*
-   * The service canonical name, or name if no canonical name
+   * The service canonical name, or name if no canonical name.
    */
   private String serviceCanonicalName;
 
   /*
-   * The service version string
+   * The service version string.
    */
   private String serviceVersion;
 
+  private Map<String, AuthType> authOverrides = new HashMap<>();
+
+  /*
+   * If present in the scope list, indicates that the API supports application default credentials
+   * based auth.
+   */
+  private static final String CLOUD_PLATFORM_SCOPE =
+      "https://www.googleapis.com/auth/cloud-platform";
+
+  /*
+   * Possible auth types supported by discovery.
+   */
+  public enum AuthType {
+    APPLICATION_DEFAULT_CREDENTIALS,
+    OAUTH_3L,
+    API_KEY
+  }
+
+  /*
+   * Returns the auth type supported by the service.
+   */
+  public AuthType getAuthType() {
+    String key = getServiceCanonicalName();
+    if (authOverrides.containsKey(key)) {
+      return authOverrides.get(key);
+    }
+    // This statement is based on the assumption that every method in a service contains all the
```

Comment:
To clarify: you're saying that we look at the scopes to determine the type of template we use for all methods (3LO, ADC, API Key), and, if we use 3LO, for each method we only include in the snippet the scopes listed for that method? That sounds OK.  I was incorrectly assuming you would use for all methods the scopes you read from the first method.

Wrt unauthenticated calls: It would be good to mark them as unauthenticated so that it's obvious to the user. I assume with 3LO this can happen when we notice no scopes listed for the method. For API Keys, I'm not sure what form this can take in Discovery, so maybe we still pass the API Key along unnecessarily (but harmlessly). Thoughts?